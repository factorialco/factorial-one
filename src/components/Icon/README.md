# Icon Component

The Icon component is used to render SVG icons from the F0 Design System in React Native applications.

## Import Pattern

```tsx
// Import the Icon component
import { Icon } from "@factorialco/f0-react-native";

// Import specific icons from app or modules directories
import { AppIcons, ModuleIcons } from "@factorialco/f0-react-native";

// Use the Icon component with the icon as a prop
<Icon icon={AppIcons.Archive} size="md" />
<Icon icon={ModuleIcons.Home} size="lg" />
```

## Props

| Prop      | Type                         | Default | Description                             |
| --------- | ---------------------------- | ------- | --------------------------------------- |
| icon      | IconType                     |         | The icon component to render            |
| size      | "xs" \| "sm" \| "md" \| "lg" | "md"    | The size of the icon                    |
| className | string                       |         | Additional classes for styling the icon |
| testID    | string                       |         | Test ID for testing                     |
| ...props  | SvgProps                     |         | Additional props for the SVG component  |

## Styling with NativeWind

The Icon component is designed to work with NativeWind, allowing you to style icons using Tailwind CSS classes:

```tsx
// Style using Tailwind CSS classes
<Icon icon={AppIcons.Archive} className="text-f1-icon-secondary" />
```

## Available Icons

The library includes two sets of icons:

1. **App Icons** - General purpose icons used throughout the application
2. **Module Icons** - Icons representing specific modules in the Factorial application

## Examples

```tsx
import { Icon } from "@factorialco/f0-react-native";
import { AppIcons, ModuleIcons } from "@factorialco/f0-react-native";

// Basic usage
<Icon icon={AppIcons.Calendar} />

// With size variant
<Icon icon={AppIcons.ChevronDown} size="xs" />
<Icon icon={AppIcons.Check} size="sm" />
<Icon icon={ModuleIcons.Home} size="md" />
<Icon icon={ModuleIcons.Settings} size="lg" />

// With color styling
<Icon icon={AppIcons.Heart} className="text-red-500" />
<Icon icon={AppIcons.InfoCircle} className="text-blue-500" />
```
