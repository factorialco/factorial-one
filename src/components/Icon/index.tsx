import { cva, type VariantProps } from "cva";
import React, {
  forwardRef,
  ForwardRefExoticComponent,
  RefAttributes,
} from "react";
import { SvgProps, Svg } from "react-native-svg";
import { cn } from "../../lib/utils";
import { cssInterop } from "nativewind";

const iconVariants = cva({
  base: "shrink-0",
  variants: {
    size: {
      xl: "w-8 h-8 stroke-xl",
      lg: "w-6 h-6 stroke-lg",
      md: "w-5 h-5 stroke-md",
      sm: "w-4 h-4 stroke-sm",
      xs: "w-3 h-3 stroke-xs",
    },
  },
  defaultVariants: {
    size: "md",
  },
});

export interface IconProps extends SvgProps, VariantProps<typeof iconVariants> {
  icon: IconType;
  testID?: string;
  className?: string;
  variant?:
    | "default"
    | "critical"
    | "neutral"
    | "ghost"
    | "outline"
    | "promote";
  isPressed?: boolean;
}

export type IconType = ForwardRefExoticComponent<
  SvgProps &
    RefAttributes<Svg> & {
      className?: string;
    }
>;

// Keep track of icons that have already had cssInterop applied
const interopAppliedIcons = new WeakSet();

// Function to apply NativeWind interop to an icon component
export function applyIconInterop(icon: IconType) {
  if (!interopAppliedIcons.has(icon)) {
    cssInterop(icon, {
      className: {
        target: "style",
        nativeStyleToProp: {
          color: true,
          opacity: true,
        },
      },
    });
    interopAppliedIcons.add(icon);
  }
  return icon;
}

export const Icon = forwardRef<Svg, IconProps>(function Icon(
  { size, icon, className, testID, ...props },
  ref,
) {
  if (!icon) return null;

  // Apply NativeWind interop to the icon if not already applied
  const Component = applyIconInterop(icon);

  return (
    <Component
      ref={ref}
      {...props}
      className={cn(iconVariants({ size }), className)}
      testID={testID}
    />
  );
});
