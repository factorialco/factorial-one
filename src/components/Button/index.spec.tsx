import { render, fireEvent, screen } from "@testing-library/react-native";
import React from "react";
import { Button } from "./";
import { IconType } from "../Icon";

// Mock the Icon component
jest.mock("../Icon", () => ({
  Icon: () => null,
}));

// Mock dependencies
const mockIcon: IconType = "check" as unknown as IconType;
const mockOnPress = jest.fn();

describe("Button", () => {
  const defaultProps = {
    label: "Test Button",
    onPress: mockOnPress,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Snapshot - default button", () => {
    const { toJSON } = render(<Button {...defaultProps} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - outline variant", () => {
    const { toJSON } = render(<Button {...defaultProps} variant="outline" />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - critical variant", () => {
    const { toJSON } = render(<Button {...defaultProps} variant="critical" />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - with icon", () => {
    const { toJSON } = render(<Button {...defaultProps} icon={mockIcon} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - with emoji", () => {
    const { toJSON } = render(<Button {...defaultProps} emoji="ðŸ‘‹" />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - disabled state", () => {
    const { toJSON } = render(<Button {...defaultProps} disabled />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - loading state", () => {
    const { toJSON } = render(<Button {...defaultProps} loading />);

    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - different sizes", () => {
    const sizes = ["sm", "md", "lg"] as const;

    sizes.forEach((size) => {
      const { toJSON } = render(<Button {...defaultProps} size={size} />);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("Snapshot - round button with hidden label", () => {
    const { toJSON } = render(<Button {...defaultProps} round hideLabel />);

    expect(toJSON()).toMatchSnapshot();
  });

  // Regular component tests
  it("renders correctly with required props", () => {
    render(<Button {...defaultProps} />);

    const button = screen.getByText("Test Button");

    expect(button).toBeDefined();
  });

  it("handles press events", () => {
    render(<Button {...defaultProps} />);

    fireEvent.press(screen.getByRole("button"));

    expect(mockOnPress).toHaveBeenCalled();
  });

  it("does not call onPress when disabled", () => {
    render(<Button {...defaultProps} disabled />);

    fireEvent.press(screen.getByRole("button"));

    expect(mockOnPress).not.toHaveBeenCalled();
  });

  it("shows correct accessibility label when disabled", () => {
    render(<Button {...defaultProps} disabled />);

    const button = screen.getByRole("button");

    expect(button.props.accessibilityLabel).toBe("Test Button, disabled");
  });

  it("shows correct accessibility label when loading", () => {
    render(<Button {...defaultProps} loading />);

    const button = screen.getByRole("button");

    expect(button.props.accessibilityLabel).toBe(
      "Test Button, disabled, loading",
    );
  });
});
