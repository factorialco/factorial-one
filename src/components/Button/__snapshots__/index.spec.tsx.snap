// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot - critical variant 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-secondary border border-f1-border h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-critical"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - default button 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - different sizes 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-6 rounded-sm gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - different sizes 2`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - different sizes 3`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-10 rounded-md gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - disabled state 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button, disabled"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded opacity-50 gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - loading state 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button, disabled, loading"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": true,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded opacity-50 gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - outline variant 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-inverse-secondary border border-f1-border h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - round button with hidden label 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded aspect-square p-0"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  />
</View>
`;

exports[`Button Snapshot - with emoji 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      👋
    </Text>
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;

exports[`Button Snapshot - with icon 1`] = `
<View
  className="flex item-start"
>
  <View
    accessibilityLabel="Test Button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="flex-row items-center justify-center border-none grow-0 bg-f1-background-accent-bold h-8 rounded gap-1 px-2 sm:px-3 lg:px-4"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      className="text-base font-medium text-f1-foreground-inverse"
    >
      Test Button
    </Text>
  </View>
</View>
`;
